File: app/src/main/java/com/boardgamegeek/ui/model/PlayModel.java
package com.boardgamegeek.ui.model;

import android.content.Context;
import android.database.Cursor;

import com.boardgamegeek.provider.BggContract.Games;
import com.boardgamegeek.provider.BggContract.Plays;
import com.boardgamegeek.util.CursorUtils;

public class PlayModel {
	public static final String[] PROJECTION = {
		Plays._ID,
		Plays.PLAY_ID,
		Plays.DATE,
		Plays.ITEM_NAME,
		Plays.OBJECT_ID,
		Plays.LOCATION,
		Plays.QUANTITY,
		Plays.LENGTH,
		Plays.PLAYER_COUNT,
		Games.THUMBNAIL_URL,
		Games.IMAGE_URL,
		Plays.COMMENTS,
		Plays.DELETE_TIMESTAMP,
		Plays.UPDATE_TIMESTAMP,
		Plays.DIRTY_TIMESTAMP
	};

	private static final int PLAY_ID = 1;
	private static final int DATE = 2;
	private static final int GAME_NAME = 3;
	private static final int GAME_ID = 4;
	private static final int LOCATION = 5;
	private static final int QUANTITY = 6;
	private static final int LENGTH = 7;
	private static final int PLAYER_COUNT = 8;
	private static final int THUMBNAIL_URL = 9;
	private static final int IMAGE_URL = 10;
	private static final int COMMENTS = 11;
	private static final int DELETE_TIMESTAMP = 12;
	private static final int UPDATE_TIMESTAMP = 13;
	private static final int DIRTY_TIMESTAMP = 14;

	private int playId;
	private int gameId;
	private String name;
	private String date;
	private String location;
	private int quantity;
	private int length;
	private int playerCount;
	private String comments;
	private String thumbnailUrl;
	private String imageUrl;
	private long deleteTimestamp;
	private long updateTimestamp;
	private long dirtyTimestamp;

	public static PlayModel fromCursor(Cursor cursor, Context context) {
		PlayModel play = new PlayModel();
		play.playId = cursor.getInt(PLAY_ID);
		play.name = cursor.getString(GAME_NAME);
		play.gameId = cursor.getInt(GAME_ID);
		play.date = CursorUtils.getFormattedDateAbbreviated(cursor, context, DATE);
		play.location = cursor.getString(LOCATION);
		play.quantity = cursor.getInt(QUANTITY);
		play.length = cursor.getInt(LENGTH);
		play.playerCount = cursor.getInt(PLAYER_COUNT);
		play.comments = CursorUtils.getString(cursor, COMMENTS).trim();
		play.thumbnailUrl = cursor.getString(THUMBNAIL_URL);
		play.imageUrl = cursor.getString(IMAGE_URL);
		play.deleteTimestamp = cursor.getLong(DELETE_TIMESTAMP);
		play.updateTimestamp = cursor.getLong(UPDATE_TIMESTAMP);
		play.dirtyTimestamp = cursor.getLong(DIRTY_TIMESTAMP);
		return play;
	}

	public int getPlayId() {
		return playId;
	}

	public String getName() {
		return name;
	}

	public String getDate() {
		return date;
	}

	public String getLocation() {
		return location;
	}

	public int getQuantity() {
		return quantity;
	}

	public int getLength() {
		return length;
	}

	public int getPlayerCount() {
		return playerCount;
	}

	public String getComments() {
		return comments;
	}

	public int getGameId() {
		return gameId;
	}

	public String getThumbnailUrl() {
		return thumbnailUrl;
	}

	public String getImageUrl() {
		return imageUrl;
	}

	public long getDeleteTimestamp() {
		return deleteTimestamp;
	}

	public long getUpdateTimestamp() {
		return updateTimestamp;
	}

	public long getDirtyTimestamp() {
		return dirtyTimestamp;
	}
}

File: app/src/main/java/com/boardgamegeek/ui/model/PlayModel.kt
package com.boardgamegeek.ui.model

import android.content.Context
import android.database.Cursor
import android.database.CursorIndexOutOfBoundsException
import android.text.format.DateUtils
import com.boardgamegeek.provider.BggContract
import com.boardgamegeek.provider.BggContract.Games
import com.boardgamegeek.provider.BggContract.Plays
import timber.log.Timber
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class PlayModel(
        private val context: Context,
        val internalId: Long,
        val playId: Int,
        val gameId: Int,
        val name: String = "",
        private val rawDate: String = "",
        val location: String = "",
        val quantity: Int = 1,
        val length: Int = 0,
        val playerCount: Int = 0,
        val comments: String = "",
        val thumbnailUrl: String = "",
        val imageUrl: String = "",
        val heroImageUrl: String = "",
        val deleteTimestamp: Long = 0,
        val updateTimestamp: Long = 0,
        val dirtyTimestamp: Long = 0
) {

    private constructor(context: Context) : this(context, BggContract.INVALID_ID.toLong(), BggContract.INVALID_ID, BggContract.INVALID_ID)

    val date: String by lazy {
        if (dateInMillis == UNKNOWN_DATE)
            ""
        else
            DateUtils.formatDateTime(context, dateInMillis, DateUtils.FORMAT_SHOW_WEEKDAY or  DateUtils.FORMAT_SHOW_DATE or DateUtils.FORMAT_ABBREV_ALL)
    }

    val dateInMillis: Long by lazy {
        if (!rawDate.isEmpty()) {
            try {
                return@lazy FORMAT.parse(rawDate).time
            } catch (e: ParseException) {
                Timber.w(e, "Unable to parse %s as %s", rawDate, FORMAT)
            } catch (e: ArrayIndexOutOfBoundsException) {
                Timber.w(e, "Unable to parse %s as %s", rawDate, FORMAT)
            }
        }
        UNKNOWN_DATE
    }

    companion object {
        const val UNKNOWN_DATE: Long = -1
        private val FORMAT = SimpleDateFormat("yyyy-MM-dd", Locale.US)

        @JvmStatic
        val projection = arrayOf(
                Plays._ID,
                Plays.PLAY_ID,
                Plays.DATE,
                Plays.ITEM_NAME,
                Plays.OBJECT_ID,
                Plays.LOCATION,
                Plays.QUANTITY,
                Plays.LENGTH,
                Plays.PLAYER_COUNT,
                Games.THUMBNAIL_URL,
                Games.IMAGE_URL,
                Plays.COMMENTS,
                Plays.DELETE_TIMESTAMP,
                Plays.UPDATE_TIMESTAMP,
                Plays.DIRTY_TIMESTAMP,
                Games.HERO_IMAGE_URL
        )

        private const val INTERNAL_ID = 0
        private const val PLAY_ID = 1
        private const val DATE = 2
        private const val GAME_NAME = 3
        private const val GAME_ID = 4
        private const val LOCATION = 5
        private const val QUANTITY = 6
        private const val LENGTH = 7
        private const val PLAYER_COUNT = 8
        private const val THUMBNAIL_URL = 9
        private const val IMAGE_URL = 10
        private const val COMMENTS = 11
        private const val DELETE_TIMESTAMP = 12
        private const val UPDATE_TIMESTAMP = 13
        private const val DIRTY_TIMESTAMP = 14
        private const val HERO_IMAGE_URL = 15

        @JvmStatic
        fun fromCursor(cursor: Cursor, context: Context): PlayModel {
            try {
                return PlayModel(
                        context,
                        cursor.getLong(INTERNAL_ID),
                        cursor.getInt(PLAY_ID),
                        cursor.getInt(GAME_ID),
                        cursor.getString(GAME_NAME) ?: "",
                        cursor.getString(DATE) ?: "",
                        cursor.getString(LOCATION) ?: "",
                        cursor.getInt(QUANTITY),
                        cursor.getInt(LENGTH),
                        cursor.getInt(PLAYER_COUNT),
                        (cursor.getString(COMMENTS) ?: "").trim(),
                        cursor.getString(THUMBNAIL_URL) ?: "",
                        cursor.getString(IMAGE_URL) ?: "",
                        cursor.getString(HERO_IMAGE_URL) ?: "",
                        cursor.getLong(DELETE_TIMESTAMP),
                        cursor.getLong(UPDATE_TIMESTAMP),
                        cursor.getLong(DIRTY_TIMESTAMP)
                )
            } catch (e: CursorIndexOutOfBoundsException) {
                Timber.w(e)
                return PlayModel(context)
            }
        }
    }
}
